### PR TITLE
[ONNX] Extend ONNX Frontend with `com.microsoft.Pad`

### LINKED ISSUE #17576: Extend ONNX Frontend with `com.microsoft.Pad` operator
## Context
Neural networks are graphs consisting of nodes called operators. Each operator corresponds to a mathematical function, usually described in framework's documentation or an AI standard, such as ONNX.

OpenVINO ONNX Frontend is a component responsible for working with ONNX graphs and requires implementation of different ONNX operators in order to use ONNX models.

This task requires extending OpenVINO ONNX Frontend with `com.microsoft.Pad` operator.

Necessary help will be provided by ONNX Fronted team.

## Operator specification
Operator details can be found in [official ONNX Runtime docs](https://github.com/microsoft/onnxruntime/blob/main/docs/ContribOperators.md#com.microsoft.Pad).

## Todo list
1. Create `.hpp` and `.cpp` files for `com.microsoft.Pad` [here](https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/onnx/frontend/src/op)
2. Prepare an implementation of this operator in form of a function. It should be placed in [opset 1 namespace](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/unique.hpp#L13). You can use the implementation of standard [ONNX Pad operator](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/op/pad.cpp) as a reference
3. Register the function in [ops_bridge.cpp](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/frontend/src/ops_bridge.cpp#L554-L560) while keeping alphabetical order
4. Create test model(s) in [ONNX models directory](https://github.com/openvinotoolkit/openvino/tree/master/src/frontends/onnx/tests/models/com.microsoft). OpenVINO test infrastructure then converts prototxt files to ONNX models - you will use those models later in tests
5. Add tests covering all use cases [here](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/tests/onnx_import_com_microsoft.in.cpp)

More details in [adding operators to ONNX Frontend guide](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/how_to_add_op.md)

## Resources

- [Contribution guide - start here!](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [Intel DevHub Discord channel](https://discord.gg/wPuqAujS) - engage in discussions, ask questions and talk to OpenVINO developers
- Guides on [ONNX Frontend architecture](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/architecture.md) and [ONNX Frontend tests](https://github.com/openvinotoolkit/openvino/blob/master/src/frontends/onnx/docs/tests.md) are also available
- You can also see a [blog post on contributing to OpenVINO](https://medium.com/openvino-toolkit/how-to-contribute-to-an-ai-open-source-project-c741f48e009e)

## Example PRs
- https://github.com/openvinotoolkit/openvino/pull/15076
- https://github.com/openvinotoolkit/openvino/pull/14936
- https://github.com/openvinotoolkit/openvino/pull/12510

## Contact points
@tomdol
@mbencer
@p-wysocki
Don't hesitate to reach out, we're here to help!

### PR DESCRIPTION
**Pull request by siddhant-0707**
### Details:
 - created implementation for `com.microsoft.Pad`
 - creating test

### Tickets:
 - Closes #17576

### CODE REVIEW

File: src/frontends/onnx/frontend/src/op/com.microsoft/pad.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        }
+    } else if (pad_mode == "edge") {
+        mode = ov::op::PadMode::EDGE;
+    } else if (pad_mode == "reflect") {
+        mode = ov::op::PadMode::REFLECT;
+    } else {
+        throw ngraph_error("Unsupported pad_mode in ONNX com.microsoft.Pad operator");
+    }
+
+    auto pads_shape = pads.get_shape();
```
Comment (mitruska) [inline src/frontends/onnx/frontend/src/op/com.microsoft/pad.cpp:39]: The `.get_shape()` will throw for dynamic shape, it should be used only if static shape is ensured, 
`.get_partial_shape()` can be used instead. 
For partial shape `.is_static()` or `.is_dynamic()` can be used and `.size()` is safe only for shapes with static rank. 
To check whether the rank is static: `partial_shape.rank().is_static()`.

**Review by mitruska — CHANGES_REQUESTED** [review]
Hello @siddhant-0707, thank you for your contribution!

[MS ONNX Pad](https://onnx.ai/onnx/operators/onnx__Pad.html#pad-11) and [ONNX standard Pad-11](https://onnx.ai/onnx/operators/onnx__Pad.html#pad-11) seem to be compatible. The only difference I can see is that MS Pad allows pads input to be `2D`, then the first dim is required to be `1`, so it could be simply resolved by Squeeze. My suggestion is to reuse `set_11::pad` code within a common function to avoid duplication of the main logic. Squeeze can be applied to the MS pads input if needed, before the common part.

Also this PR doesn't have any tests for the provided changes, please add them including the mentioned 2D pads scenario.
Registration of the correct pad function should be ensured.

**Comment by mlukasze** [comment]
hey @siddhant-0707 :)
will you continue a work on this PR? Or should we move a task back to the open pool?

**Comment by siddhant-0707** [comment]
Apologies for the delay @mlukasze, I was a little occupied with other tasks. Please reopen the PR.
@mitruska I have made changes as per your requirements and also added a test for `pads` with 2d shape. However, I'm still stuck as the `.prototxt` model doesn't get converted to `.onnx` when compiling and hence test fails.

**Comment by siddhant-0707** [comment]
Apologies for the delay @mlukasze, I was a little occupied with other tasks. Please reopen the PR.
@mitruska I have reused the implementation of `set_11::pad` (with modification in beginning to handle case with 2d `pads` input). I've done this because we have to assign the squeezed value back to `node.get_ng_inputs()[1]` (which we can't do) Is there a way to create a new node and set its inputs and pass it to `set_11::pad`?.
Another way this problem can be solved is creating an overload in `onnx::pad` that takes in `pads` separately and then passes it to `set_11::pad`. Something like `ov::OutputVector pad(const Node& node, const ov::Output<ov::Node>& pads) {}`. However, this would misalign code with docs I think.

**Comment by siddhant-0707** [comment]
Also added 2d and 1d tests

**Comment by mitruska** [comment]
Hello @siddhant-0707, great to see some updates! Reopened the PR as you asked.
To continue review, please push commits with the changes you have mentioned and sync branch with the latest master.

**Review by gkrivor — APPROVED** [review]

**Comment by gkrivor** [comment]
build_jenkins

File: src/frontends/onnx/frontend/src/op/com.microsoft/pad.cpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+namespace custom {
+namespace set_1 {
+ov::OutputVector pad(const ov::frontend::onnx::Node& node) {
+    const auto inputs = node.get_ov_inputs();
+    const auto& data = inputs[0];
+    const auto& pads_input = inputs[1];
+    auto pads = pads_input;
+    if (pads.get_shape().size() == 2) {
+        pads = std::make_shared<v0::Squeeze>(pads);
+    }
```
Comment (mitruska) [inline src/frontends/onnx/frontend/src/op/com.microsoft/pad.cpp:49]: As I mentioned, `get_shape()` called on the node with dynamic shape will cause an issue, so it should be ensured that the shape is static before the rank size check.

```suggestion
    const auto& pads_pshape = pads.get_partial_shape();
    if (pads_pshape.rank().is_static() && pads_pshape.size() == 2) {
        pads = std::make_shared<v0::Squeeze>(pads);
    }
```

**Review by mitruska — COMMENTED** [review]
My previous concern (https://github.com/openvinotoolkit/openvino/pull/22000#discussion_r1492629220) about `get_shape()` call has been not applied.
Please add suggested change to avoid further issues.

About creating the common pad function, it could be a helper taking `ov::OutputVector` as argument, and called from the onnx standard pad and custom pad (after get_ov_inputs()). 
But such code unification can be considered as follow up improvement within a separate PR. 

Let's focus on the last necessary change:)

**Review by mitruska — APPROVED** [review]

**Comment by mitruska** [comment]
build_jenkins

**Comment by mitruska** [comment]
build_jenkins

**Pull request merged**